okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java

```diff
@@ -982,144 +752,186 @@ void listModulesFromInterface(String tenantId,
-      for (String id : modsEnabled.keySet()) {
-        ModuleId moduleId = new ModuleId(id);
-        String latestId = moduleId.getLatest(modsAvailable.keySet());
-        if (!latestId.equals(id)) {
-          TenantModuleDescriptor tmd = new TenantModuleDescriptor();
-          tmd.setAction(Action.enable);
-          tmd.setId(latestId);
-          logger.info("upgrade.. enable {}", latestId);
-          tmd.setFrom(id);
-          tml2.add(tmd);
-        }
+  private List<TenantModuleDescriptor> upgrades(
+      Map<String, ModuleDescriptor> modsAvailable, Map<String, ModuleDescriptor> modsEnabled) {
+
+    List<TenantModuleDescriptor> tml = new LinkedList<>();
+    for (String id : modsEnabled.keySet()) {
+      ModuleId moduleId = new ModuleId(id);
+      String latestId = moduleId.getLatest(modsAvailable.keySet());
+      if (!latestId.equals(id)) {
+        TenantModuleDescriptor tmd = new TenantModuleDescriptor();
+        tmd.setAction(Action.enable);
+        tmd.setId(latestId);
+        logger.info("upgrade.. enable {}", latestId);
+        tmd.setFrom(id);
+        tml.add(tmd);
       }
-      return tml2;
-    } else {
-      return tml;
     }
+    return tml;
   }
 
-  private void installUpgradeModules2(
-      Tenant t, ProxyContext pc,
-      TenantInstallOptions options,
+  private Future<List<TenantModuleDescriptor>> runJob(
+      Tenant t, ProxyContext pc, TenantInstallOptions options,
       Map<String, ModuleDescriptor> modsAvailable,
-      Map<String, ModuleDescriptor> modsEnabled, List<TenantModuleDescriptor> tml,
-      Handler<ExtendedAsyncResult<List<TenantModuleDescriptor>>> fut) {
+      Map<String, ModuleDescriptor> modsEnabled, InstallJob job) {
 
-    DepResolution.installSimulate(modsAvailable, modsEnabled, tml, res -> {
-      if (res.failed()) {
-        fut.handle(new Failure<>(res.getType(), res.cause()));
-        return;
-      }
+    List<TenantModuleDescriptor> tml = job.getModules();
+    return DepResolution.installSimulate(modsAvailable, modsEnabled, tml).compose(res -> {
       if (options.getSimulate()) {
-        fut.handle(new Success<>(tml));
-      } else {
-        installAutodeploy(t, pc, options, modsAvailable, tml, tml.iterator(),
-            res1 -> {
-              if (res1.failed()) {
-                fut.handle(new Failure<>(res1.getType(), res1.cause()));
-              } else {
-                fut.handle(new Success<>(tml));
+        return Future.succeededFuture(tml);
+      }
+      return jobs.add(t.getId(), job.getId(), job).compose(res2 -> {
+        Promise<List<TenantModuleDescriptor>> promise = Promise.promise();
+        Future<Void> future = Future.succeededFuture();
+        if (options.getAsync()) {
+          List<TenantModuleDescriptor> tml2 = new LinkedList<>();
+          for (TenantModuleDescriptor tm : tml) {
+            tml2.add(tm.cloneWithoutStatus());
+          }
+          promise.complete(tml2);
```

Comment:
Should we also set stage to be `pending` for `tm` in `tml2`?

Reply:
No. That wasn't the intention. The meaning was that it would return entries without status/stage .. for the async immediate response.